File: sorting/radix_example.go
package main

import (
	"bytes"
	"encoding/binary"
	"fmt"
)

const digit = 4
const maxbit = -1 << 31

// Constants are already defined in radix_sort.go

func mainRad() {
	// Simple example with mixed positive/negative numbers
	data := []int32{-2, 1, -1, 2}
	fmt.Printf("Original data: %v\n\n", data)

	radixSortWithTrace(data)

	fmt.Printf("\nFinal sorted result: %v\n", data)
}

func radixSortWithTrace(data []int32) {
	fmt.Println("=== STEP 1: Convert to bytes with sign bit flip ===")

	buf := bytes.NewBuffer(nil)
	ds := make([][]byte, len(data))

	for i, e := range data {
		fmt.Printf("Original: %d (binary: %032b)\n", e, uint32(e))
		fmt.Printf("%b", maxbit)
		//fmt.Sprintf("%b", e)
		flipped := e ^ maxbit
		fmt.Printf("After XOR:  %d (binary: %032b)\n", flipped, uint32(flipped))

		binary.Write(buf, binary.LittleEndian, flipped)
		b := make([]byte, digit)
		buf.Read(b)
		ds[i] = b

		fmt.Printf("As bytes: [%d, %d, %d, %d] (little-endian)\n", b[0], b[1], b[2], b[3])
		fmt.Println()
	}

	fmt.Println("=== STEP 2: Radix sort by each byte position ===")

	countingSort := make([][][]byte, 256)

	for bytePos := 0; bytePos < digit; bytePos++ {
		fmt.Printf("\n--- Processing byte position %d ---\n", bytePos)

		// Show current state
		fmt.Print("Current order: ")
		for i, b := range ds {
			if i > 0 {
				fmt.Print(", ")
			}
			fmt.Printf("[%d,%d,%d,%d]", b[0], b[1], b[2], b[3])
		}
		fmt.Println()

		// Distribute into buckets
		fmt.Printf("Sorting by byte[%d]:\n", bytePos)
		for i, b := range ds {
			bucketIndex := b[bytePos]
			countingSort[bucketIndex] = append(countingSort[bucketIndex], b)
			fmt.Printf("  Item %d: byte[%d]=%d -> bucket %d\n", i, bytePos, bucketIndex, bucketIndex)
		}

		// Show non-empty buckets
		fmt.Println("Non-empty buckets:")
		for k := 0; k < 256; k++ {
			if len(countingSort[k]) > 0 {
				fmt.Printf("  Bucket %d: ", k)
				for i, b := range countingSort[k] {
					if i > 0 {
						fmt.Print(", ")
					}
					fmt.Printf("[%d,%d,%d,%d]", b[0], b[1], b[2], b[3])
				}
				fmt.Println()
			}
		}

		// Collect back in order
		j := 0
		for k, bs := range countingSort {
			if len(bs) > 0 {
				copy(ds[j:], bs)
				j += len(bs)
				countingSort[k] = bs[:0]
			}
		}

		// Show result after this pass
		fmt.Print("After sorting: ")
		for i, b := range ds {
			if i > 0 {
				fmt.Print(", ")
			}
			fmt.Printf("[%d,%d,%d,%d]", b[0], b[1], b[2], b[3])
		}
		fmt.Println()
	}

	fmt.Println("\n=== STEP 3: Convert back to integers and restore sign ===")

	var w int32
	for i, b := range ds {
		buf.Write(b)
		binary.Read(buf, binary.LittleEndian, &w)
		restored := w ^ maxbit

		fmt.Printf("Bytes [%d,%d,%d,%d] -> %d -> %d (after XOR back)\n",
			b[0], b[1], b[2], b[3], w, restored)

		data[i] = restored
	}
}

File: algo/sorting/radix_example.py
import struct
from typing import List

DIGIT = 4
MAXBIT = -1 << 31  # 0x80000000

def main():
    # Simple example with mixed positive/negative numbers
    data = [-2, 1, -1, 2]
    print(f"Original data: {data}\n")
    
    radix_sort_with_trace(data)
    
    print(f"\nFinal sorted result: {data}")

def radix_sort_with_trace(data: List[int]):
    print("=== STEP 1: Convert to bytes with sign bit flip ===")
    
    ds = []  # List of byte arrays
    
    for i, e in enumerate(data):
        print(f"Original: {e} (binary: {e & 0xFFFFFFFF:032b})")
        flipped = e ^ MAXBIT
        print(f"After XOR:  {flipped} (binary: {flipped & 0xFFFFFFFF:032b})")
        
        # Convert to little-endian bytes
        b = list(struct.pack('<i', flipped))  # '<i' = little-endian signed int
        ds.append(b)
        
        print(f"As bytes: {b} (little-endian)")
        print()
    
    print("=== STEP 2: Radix sort by each byte position ===")
    
    for byte_pos in range(DIGIT):
        print(f"\n--- Processing byte position {byte_pos} ---")
        
        # Show current state
        print("Current order: ", end="")
        for i, b in enumerate(ds):
            if i > 0:
                print(", ", end="")
            print(f"{b}", end="")
        print()
        
        # Create counting sort buckets (256 buckets for each possible byte value)
        counting_sort = [[] for _ in range(256)]
        
        # Distribute into buckets
        print(f"Sorting by byte[{byte_pos}]:")
        for i, b in enumerate(ds):
            # Convert signed byte to unsigned (0-255 range)
            bucket_index = b[byte_pos] if b[byte_pos] >= 0 else b[byte_pos] + 256
            counting_sort[bucket_index].append(b)
            print(f"  Item {i}: byte[{byte_pos}]={bucket_index} -> bucket {bucket_index}")
        
        # Show non-empty buckets
        print("Non-empty buckets:")
        for k in range(256):
            if len(counting_sort[k]) > 0:
                print(f"  Bucket {k}: ", end="")
                for i, b in enumerate(counting_sort[k]):
                    if i > 0:
                        print(", ", end="")
                    print(f"{b}", end="")
                print()
        
        # Collect back in order (stable sort)
        ds.clear()
        for bucket in counting_sort:
            ds.extend(bucket)
        
        # Show result after this pass
        print("After sorting: ", end="")
        for i, b in enumerate(ds):
            if i > 0:
                print(", ", end="")
            print(f"{b}", end="")
        print()
    
    print("\n=== STEP 3: Convert back to integers and restore sign ===")
    
    for i, b in enumerate(ds):
        # Convert bytes back to integer
        w = struct.unpack('<i', bytes(b))[0]  # '<i' = little-endian signed int
        restored = w ^ MAXBIT
        
        print(f"Bytes {b} -> {w} -> {restored} (after XOR back)")
        
        data[i] = restored

if __name__ == "__main__":
    main()